src/attached_probe.h

```diff
@@ -40,84 +40,208 @@ class AttachedProbe {
+private:
+  AttachedTracepointProbe(int progfd,
+                          int perf_event_fd,
+                          std::string event_name,
+                          std::string probe_path);
+  int perf_event_fd_;
+  const std::string event_name_;
+  const std::string probe_path_;
+};
+
+class AttachedProfileProbe : public AttachedProbe {
+public:
+  static Result<std::unique_ptr<AttachedProfileProbe>> make(
+      Probe &probe,
+      const BpfProgram &prog,
+      std::optional<int> pid);
+  ~AttachedProfileProbe() override;
+
+private:
+  AttachedProfileProbe(int progfd, std::vector<int> perf_event_fds);
+  std::vector<int> perf_event_fds_;
+};
+
+class AttachedIntervalProbe : public AttachedProbe {
+public:
+  static Result<std::unique_ptr<AttachedIntervalProbe>> make(
+      Probe &probe,
+      const BpfProgram &prog,
+      std::optional<int> pid);
+  ~AttachedIntervalProbe() override;
+
+private:
+  AttachedIntervalProbe(int progfd, int perf_event_fd);
+  int perf_event_fd_;
+};
+
+class AttachedSoftwareProbe : public AttachedProbe {
+public:
+  static Result<std::unique_ptr<AttachedSoftwareProbe>> make(
+      Probe &probe,
+      const BpfProgram &prog,
+      std::optional<int> pid);
+  ~AttachedSoftwareProbe() override;
+
+private:
+  AttachedSoftwareProbe(int progfd, std::vector<int> perf_event_fds);
+  std::vector<int> perf_event_fds_;
+};
+
+class AttachedHardwareProbe : public AttachedProbe {
+public:
+  static Result<std::unique_ptr<AttachedHardwareProbe>> make(
+      Probe &probe,
+      const BpfProgram &prog,
+      std::optional<int> pid);
+  ~AttachedHardwareProbe() override;
+
+private:
+  AttachedHardwareProbe(int progfd, std::vector<int> perf_event_fds);
+  std::vector<int> perf_event_fds_;
+};
+
+class AttachedFentryProbe : public AttachedProbe {
+public:
+  static Result<std::unique_ptr<AttachedFentryProbe>> make(
+      const BpfProgram &prog);
+  ~AttachedFentryProbe() override;
+
+private:
+  AttachedFentryProbe(int progfd, int tracing_fd);
+  int tracing_fd_;
+};
+
+class AttachedRawtracepointProbe : public AttachedProbe {
+public:
+  static Result<std::unique_ptr<AttachedRawtracepointProbe>> make(
+      Probe &probe,
+      const BpfProgram &prog);
+  ~AttachedRawtracepointProbe() override;
+
+private:
+  AttachedRawtracepointProbe(int progfd, int tracing_fd);
+  int tracing_fd_;
+};
+
+class AttachedIterProbe : public AttachedProbe {
+public:
+  static Result<std::unique_ptr<AttachedIterProbe>> make(
+      const BpfProgram &prog,
+      std::optional<int> pid);
+  ~AttachedIterProbe() override;
+
+  const int iter_link_fd;
+
+private:
+  AttachedIterProbe(int progfd, int iter_link_fd);
+};
+
+class AttachedWatchpointProbe : public AttachedProbe {
+public:
+  static Result<std::unique_ptr<AttachedWatchpointProbe>> make(
+      Probe &probe,
+      const BpfProgram &prog,
+      std::optional<int> pid,
+      const std::string &mode);
+  ~AttachedWatchpointProbe() override;
+
+  const uint64_t probe_addr;
```

Comment:
Instead of divergent subclasses, how about wiring this through a virtual getter on the base class? On the base class, you could supply a default EOPNOTSUPP (some equivalent, of course) implementation. Then relevant subclasses can override. That way callers don't need to dynamic cast and we can hide all these class definitions in attached_probe.cpp and only expose `class AttachedProbe` in `attached_probe.h`.

Also, it might make more sense to just add a `get_probe()` that returns `const Probe&`?

Reply:
> Also, it might make more sense to just add a get_probe() that returns const Probe&?

I think I prefer this. Will update.